text/0000-closure-to-fn-coercion.md

```diff
@@ -0,0 +1,222 @@
+  &|var: &mut u32| {},
+  &|var: &mut u32| *var += 1,
+  &|var: &mut u32| *var += 2,
+  &|var: &mut u32| *var += 3,
+];
+```
+
+And this seems to fix the problem. Unfortunately, however, because we use
+a reference to the `Fn` trait, an extra layer of indirection is added when
+attempting to run `foo[n](&mut bar)`.
+
+Rust must use dynamic dispatch in this situation; a closure with captures is nothing
+but a struct containing references to captured variables. The code associated with a
+closure must be able to access those references stored in the struct.
+
+In situations where this function pointer array is particularly hot code,
+any optimizations would be appreciated. More generally, it is always preferable
+to avoid unnecessary indirection. And, of course, it is impossible to use this syntax
+when dealing with FFI.
+
+Aside from code-size nits, anonymous functions are legitimately useful for programmers.
+In the case of callback-heavy code, for example, it can be impractical to define functions
+out-of-line, with the requirement of producing confusing (and unnecessary) names for each.
+In the very first example given, `inc_X` names were used for the out-of-line functions, but
+more complicated behavior might not be so easily representable.
+
+Finally, this sort of automatic coercion is simply intuitive to the programmer.
+In the `&Fn` example, no variables are captured by the closures, so the theory is
+that nothing stops the compiler from treating them as anonymous functions.
+
+# Detailed design
+[design]: #detailed-design
+
+In C++, non-capturing lambdas (the C++ equivalent of closures) "decay" into function pointers
+when they do not need to capture any variables. This is used, for example, to pass a lambda
+into a C function:
+
+```cpp
+void foo(void (*foobar)(void)) {
+    // impl
+}
+void bar() {
+    foo([]() { /* do something */ });
+}
+```
+
+With this proposal, rust users would be able to do the same:
+
+```rust
+fn foo(foobar: fn()) {
+    // impl
+}
+fn bar() {
+    foo(|| { /* do something */ });
+}
+```
+
+Using the examples within ["Motivation"](#motivation), the code array would
+be simplified to no performance detriment:
+
+```rust
+const foo: [fn(&mut u32); 4] = [
+  |var: &mut u32| {},
+  |var: &mut u32| *var += 1,
+  |var: &mut u32| *var += 2,
+  |var: &mut u32| *var += 3,
+];
+```
+
+Because there does not exist any item in the language that directly produces
+a `fn` type, even `fn` items must go through the process of reification. To
+perform the coercion, then, rustc must additionally allow the reification of
+unsized closures to `fn` types. The implementation of this is simplified by the
+fact that closures' capture information is recorded on the type-level.
+
+*Note:* once explicitly assigned to an `Fn` trait, the closure can no longer be
+coerced into `fn`, even if it has no captures.
+
+```rust
+let a: &Fn(u32) -> u32 = |foo: u32| { foo + 1 };
+let b: fn(u32) -> u32 = *a; // Can't re-coerce
+```
```

Comment:
Should I make it explicit in this RFC that it describes a coercion between the closure and fn-pointer *types*, rather than an implicit conversion of closure *expressions* to anonymous function pointers?